crates/data_components/src/duckdb/write.rs

```diff
@@ -142,26 +142,41 @@ impl DataSink for DuckDBDataSink {
             .context(super::UnableToBeginTransactionSnafu)
             .map_err(to_datafusion_error)?;
 
-        let num_rows = self.try_write_all(&tx, &data_batches)?;
+        let num_rows = match **self.duckdb.constraints() {
+            [] => self.try_write_all_no_constraints(&tx, &data_batches)?,
+            _ => self.try_write_all_with_constraints(&tx, &data_batches)?,
+        };
 
         tx.commit()
-            .context(super::UnableToCommitDuckDBTransactionSnafu)
+            .context(super::UnableToCommitTransactionSnafu)
             .map_err(to_datafusion_error)?;
 
         Ok(num_rows)
     }
 }
 
 impl DuckDBDataSink {
-    fn new(duckdb: Arc<DuckDB>, overwrite: bool) -> Self {
+    pub(crate) fn new(duckdb: Arc<DuckDB>, overwrite: bool) -> Self {
         Self { duckdb, overwrite }
     }
 
-    fn try_write_all(
+    /// If there are constraints on the `DuckDB` table, we need to create an empty copy of the target table, write to that table copy and then depending on
```

Comment:
Good to have https://duckdb.org/docs/sql/indexes#over-eager-unique-constraint-checking referenced here.